src/reduce.py

```diff
@@ -13,8 +25,19 @@ def reduce(f: Callable[[A], B], x: list[A]) -> B:
     6
     """
     assert len(x) >= 2
-    ...  # FIXME
+    
+    n = len(x)
+    y = [0]*n
+    y[0] = x[0]
+    for i in range(1, len(x)):
+        y[i] = f(y[i-1], x[i])
+    return y[n-1]
```

Comment:
You do not need to store all the intermediate results here. Admittedly, some implementations of this functions do, but it is generally wasteful of memory. You could instead use a single variable to hold the intermediates

```python
    res = x[0]
    for i in range(1, len(x)):
        res = f(res, x[i])
    return res
```